### DISCUSSION TITLE
Custom operator library distribution

### DISCUSSION THREAD

**Discussion by netaz** [microsoft/onnxruntime#14181] [General] [answered]
Hello!
I have an ORT custom operator bundled in an .so file which I want to distribute to users. Users may install any of several ORT versions (depending on OS and Python version support).

The problem I'm facing is that each ORT version requires my custom operator to be compiled with that ORT version's header files (e.g. in v1.13.1 the custom op requires 4 headers, and in v1.10.0 the same code requires 3 headers).
Is there a way to handle this cleanly? For example, is there an ORT pypi distribution that installs ORT's public API headers in `site-packages`?

**Comment by xadupre** [comment]
The header are released on github (https://github.com/microsoft/onnxruntime/releases). You can download the corresponding headers during your installation before the compilation starts.

**Reply by netaz** [reply]
Thanks for the prompt reply.
That would entail cloning all of ORT, right? It's a bit cumbersome to add this to an installation process of a product in which the ORT is just a very small component, and it adds more potential for installation failure.

**Reply by xadupre** [reply]
Otherwise, you can just add the C++ headers to your sources.

**Comment by xadupre** [comment, marked as answer]
You don't have to clone, you can just download the release. In that file [setup.py](https://github.com/sdpython/mloptonnx/blob/main/setup.py#L151), I overwritten the `build_ext` command to download the release first.

**Reply by netaz** [reply]
Thanks @xadupre,
Our previous methodology was to add the ORT C++ headers to our sources (as you suggested above) but that had us chasing ORT releases.
Downloading the release and extracting the headers is a better solution for us because we can automate (write once and forget).
